### PR TITLE
Implement CtrlScaleModAdd and CModAddK bloqs for Modular Exponentiation

### PR DESCRIPTION
**Pull request by fpapa250**
Implements CtrlScaleModAdd and CModAddK bloqs. Also fixes a typechecking error with IntEffect (found thanks to mypy). For some reason it has a serialization error on CModAddK autobloqtester for split/joins even with the ignore_split_join generalizer.

### CODE REVIEW

File: qualtran/bloqs/basic_gates/z_basis.py [hunk 1]
```diff
@@ -482,7 +482,7 @@ class IntEffect(_IntVector):
         val: The register of size `bitsize` which de-allocates the value `val`.
     """
 
-    def __init__(self, val: int, bitsize: int):
+    def __init__(self, val: Union[int, sympy.Expr], bitsize: Union[int, sympy.Expr]):
```
Comment (anurudhp) [inline qualtran/bloqs/basic_gates/z_basis.py:485]: you can use the type alias `SymbolicInt` (from `qualtran.symbolics`)

File: qualtran/bloqs/mod_arithmetic/mod_addition.py [hunk 2]
```diff
@@ -32,16 +32,17 @@
     Soquet,
     SoquetT,
 )
+from qualtran._infra.bloq import DecomposeTypeError
```
Comment (mpharrigan) [inline qualtran/bloqs/mod_arithmetic/mod_addition.py:35]: `from qualtran import DecomposeTypeError`

File: qualtran/bloqs/mod_arithmetic/mod_addition.py [hunk 3]
```diff
@@ -88,8 +89,8 @@ def on_classical_vals(
         return {'x': x, 'y': (x + y) % self.mod}
 
     def build_composite_bloq(self, bb: 'BloqBuilder', x: Soquet, y: Soquet) -> Dict[str, 'SoquetT']:
-        if is_symbolic(self.bitsize):
-            raise NotImplementedError(f'symbolic decomposition is not supported for {self}')
+        if isinstance(self.bitsize, sympy.Expr):
```
Comment (mpharrigan) [inline qualtran/bloqs/mod_arithmetic/mod_addition.py:92]: why the change from `is_symbolic` for `isinstance`?

File: qualtran/bloqs/mod_arithmetic/mod_addition.py [hunk 4]
```diff
@@ -88,8 +89,8 @@ def on_classical_vals(
...
+            raise DecomposeTypeError("Cannot decompose symbolic `bitsize`.")
```
Comment (mpharrigan) [inline qualtran/bloqs/mod_arithmetic/mod_addition.py:93]: please keep the `f"{self}" somewhere`

File: qualtran/bloqs/mod_arithmetic/mod_addition.py [hunk 5]
```diff
@@ -279,12 +280,55 @@ class CModAddK(Bloq):
     def signature(self) -> 'Signature':
         return Signature([Register('ctrl', QBit()), Register('x', QUInt(self.bitsize))])
 
+    def build_composite_bloq(
+        self, bb: 'BloqBuilder', ctrl: 'Soquet', x: 'Soquet'
+    ) -> Dict[str, 'SoquetT']:
+        if isinstance(self.bitsize, sympy.Expr):
+            raise DecomposeTypeError("Cannot decompose symbolic `bitsize`.")
```
Comment (mpharrigan) [inline qualtran/bloqs/mod_arithmetic/mod_addition.py:287]: is this true? none of the registers are split in this decomposition

File: qualtran/bloqs/mod_arithmetic/mod_addition.py [hunk 6]
```diff
@@ -316,9 +360,38 @@ def signature(self) -> 'Signature':
             ]
         )
 
+    def build_composite_bloq(
+        self, bb: 'BloqBuilder', ctrl: 'Soquet', x: 'Soquet', y: 'Soquet'
+    ) -> Dict[str, 'SoquetT']:
+        if isinstance(self.bitsize, sympy.Expr):
+            raise DecomposeTypeError("Cannot decompose symbolic `bitsize`.")
```
Comment (mpharrigan) [inline qualtran/bloqs/mod_arithmetic/mod_addition.py:367]: include {self} in the message

File: qualtran/bloqs/mod_arithmetic/mod_addition.py [hunk 7]
```diff
@@ -279,12 +280,55 @@ class CModAddK(Bloq):
     def signature(self) -> 'Signature':
         return Signature([Register('ctrl', QBit()), Register('x', QUInt(self.bitsize))])
 
+    def build_composite_bloq(
```
Comment (mpharrigan) [inline qualtran/bloqs/mod_arithmetic/mod_addition.py:283]: please add a reference for this construction

File: qualtran/bloqs/mod_arithmetic/mod_addition.py [hunk 8]
```diff
@@ -316,9 +360,38 @@ def signature(self) -> 'Signature':
...
+        x = bb.join(x_split, dtype=QUInt(self.bitsize))
+
+        return {'ctrl': ctrl, 'x': x, 'y': y}
+
     def build_call_graph(self, ssa: 'SympySymbolAllocator') -> 'BloqCountDictT':
         k = ssa.new_symbol('k')
-        return {CModAddK(k=k, bitsize=self.bitsize, mod=self.mod): self.bitsize}
+        return {
+            CModAddK(k=k, bitsize=self.bitsize, mod=self.mod): self.bitsize,
+            And(): self.bitsize,
```
Comment (mpharrigan) [inline qualtran/bloqs/mod_arithmetic/mod_addition.py:392]: this changes the cost of the bloq. Do you have a reference for the present construction? We at least need a note about the difference between gidney 2019

File: qualtran/bloqs/mod_arithmetic/mod_addition.py [hunk 9]
```diff
@@ -390,6 +481,8 @@ def on_classical_vals(
     def build_composite_bloq(
         self, bb: 'BloqBuilder', ctrl, x: Soquet, y: Soquet
     ) -> Dict[str, 'SoquetT']:
+        if isinstance(self.dtype.bitsize, sympy.Expr):
+            raise DecomposeTypeError("Cannot decompose symbolic `bitsize`.")
```
Comment (mpharrigan) [inline qualtran/bloqs/mod_arithmetic/mod_addition.py:485]: include a {self}

**Review by mpharrigan — COMMENTED** [review]
first pass

File: qualtran/bloqs/mod_arithmetic/mod_addition.py (hunk 3 above)
Comment (fpapa250) [reply qualtran/bloqs/mod_arithmetic/mod_addition.py]: I thought that was the reason some tests would fail for symbolic decomposition - now realizing it was the DecomposeTypeError vs NotImplementedError.

File: qualtran/bloqs/mod_arithmetic/mod_addition.py [hunk 10]
```diff
@@ -35,13 +36,14 @@
...
-from qualtran.bloqs.basic_gates import XGate
+from qualtran.bloqs.basic_gates import IntEffect, IntState, XGate
 from qualtran.bloqs.bookkeeping import Cast
+from qualtran.bloqs.mcmt.and_bloq import And
 from qualtran.drawing import Circle, Text, TextBox, WireSymbol
 from qualtran.resource_counting import BloqCountDictT, SympySymbolAllocator
 from qualtran.resource_counting.generalizers import ignore_split_join
 from qualtran.simulation.classical_sim import ClassicalValT
-from qualtran.symbolics import is_symbolic
+from qualtran.symbolics.types import is_symbolic
```
Comment (mpharrigan) [inline qualtran/bloqs/mod_arithmetic/mod_addition.py:46]: ?

File: qualtran/bloqs/mod_arithmetic/mod_addition.py [hunk 11]
```diff
@@ -269,6 +271,15 @@ class CModAddK(Bloq):
     Registers:
         ctrl: The control bit
         x: The register to perform the in-place modular addition.
+
+    References:
```
Comment (mpharrigan) [inline qualtran/bloqs/mod_arithmetic/mod_addition.py:275]: refactor from (refs)x(commentary) to zip(refs, commentary)

```
[arithmetic](...)
Author. yyy.

[rsa](...).
Gidney and Ekera 2019. 
Uses CModAddK in the "Reference implementation" using ref 1, fig 4 etc etc

[ecc](...).
Litinski et al. yyyy.
Uses 2 fewer add operations because K is classical (??)
```

File: qualtran/bloqs/mod_arithmetic/mod_addition.py [hunk 12]
```diff
@@ -269,6 +271,15 @@ class CModAddK(Bloq):
...
+        [How to factor 2048 bit RSA integers in 8 hours using 20 million noisy qubits](https://arxiv.org/abs/1905.09749).
+        [Quantum Networks for Elementary Arithmetic Operations](https://arxiv.org/abs/quant-ph/9511018).
+        [How to compute a 256-bit elliptic curve private key with only 50 million Toffoli gates](https://arxiv.org/abs/2306.08585).
+        Construction in the Gidney and Ekerå paper from 2019 references the second source's
+        construction in figure 4. Because the construction is simply a ModAdd bloq and not a
+        ModAddK bloq, we choose to use the bloq described in the third paper as it uses 2 fewer Add
```
Comment (mpharrigan) [inline qualtran/bloqs/mod_arithmetic/mod_addition.py:281]: syntax of contrasting ModAdd vs ModAddK is kindof confusing, but might be fixed if you split up the references per my other comment

File: qualtran/bloqs/mod_arithmetic/mod_addition.py [hunk 13]
```diff
@@ -300,6 +352,12 @@ class CtrlScaleModAdd(Bloq):
         ctrl: The control bit
         x: The 'source' quantum register containing the integer to be scaled and added to `y`.
         y: The 'destination' quantum register to which the addition will apply.
+
+    References:
+        [How to factor 2048 bit RSA integers in 8 hours using 20 million noisy qubits](https://arxiv.org/abs/1905.09749).
+        Construction based on description in section 2.2 paragraph 4. We add n And/And† bloqs
+        because the bloq is controlled, but the construction also involves modular addition
+        controlled on the qubits comprising register x.
```
Comment (mpharrigan) [inline qualtran/bloqs/mod_arithmetic/mod_addition.py:360]: I don't understand what this is trying to say. Please try to disambiguate between the construction in Qualtran and the construction in the reference

**Review by mpharrigan — COMMENTED** [review]
some wording confusion in the docstrings

File: qualtran/bloqs/mod_arithmetic/mod_addition.py (hunk 13 above)
Comment (fpapa250) [reply qualtran/bloqs/mod_arithmetic/mod_addition.py:368]: I tried to be more clear. Basically, the reference implementation uses CModAddK, but the circuit that it points to is just ModAdd (not ModAddK). This ModAdd is less efficient than the circuit later introduced in the Litinski paper so we choose to use that since it is more efficient and already implemented in Qualtran.

File: qualtran/bloqs/basic_gates/z_basis.py [hunk 14]
```diff
@@ -42,6 +42,7 @@
 )
 from qualtran.bloqs.bookkeeping import ArbitraryClifford
 from qualtran.drawing import Circle, directional_text_box, Text, TextBox, WireSymbol
+from qualtran.symbolics.types import SymbolicInt
```
Comment (mpharrigan) [inline qualtran/bloqs/basic_gates/z_basis.py:45]: just import from `qualtran.symbolics`

File: qualtran/bloqs/mod_arithmetic/mod_addition.py [hunk 15]
```diff
@@ -269,6 +271,23 @@ class CModAddK(Bloq):
     Registers:
         ctrl: The control bit
         x: The register to perform the in-place modular addition.
+
+    References:
```
Comment (mpharrigan) [inline qualtran/bloqs/mod_arithmetic/mod_addition.py:275]: this is very helpful, thanks

File: qualtran/bloqs/mod_arithmetic/mod_addition.py [hunk 16]
```diff
@@ -279,12 +298,53 @@ class CModAddK(Bloq):
...
-        return f'x += {self.k} % {self.mod}'
+    def wire_symbol(
+        self, reg: Optional['Register'], idx: Tuple[int, ...] = tuple()
+    ) -> 'WireSymbol':
+        if reg is None:
+            return Text(f"mod {self.mod}")
+        if reg.name == 'ctrl':
+            return Circle()
+        if reg.name == 'x':
+            return TextBox(f'x += {self.k} mod {self.mod}')
```
Comment (mpharrigan) [inline qualtran/bloqs/mod_arithmetic/mod_addition.py:330]: I would suggest only putting the mod value in one place. These symbols are used in the "musical score" circuit diagram where width is at a premium. The `reg is None` case puts a little label _above_ the box. The register-specific ones put text _in_ the box. I'd ommit the mod from the x box

File: qualtran/bloqs/mod_arithmetic/mod_addition.py (hunk 13 above)
Comment (mpharrigan) [reply qualtran/bloqs/mod_arithmetic/mod_addition.py:368]: can you just write that instead?

This is my secret trick for good technical writing: write all your docstrings as if you're explaining something to an actual human reading your code; because spoiler alert: that's what you're doing!

**Review by mpharrigan — COMMENTED** [review]
some final nits, then lgtm

File: qualtran/bloqs/mod_arithmetic/mod_addition.py (hunk 13 above)
Comment (fpapa250) [reply qualtran/bloqs/mod_arithmetic/mod_addition.py:364]: Simplified further and used above explanation.

**Review by mpharrigan — APPROVED** [review]
